baseline/pytorch/tagger/model.py

```diff
@@ -136,83 +129,358 @@ def make_input(self, batch_dict, perm=False, numpy_to_tensor=False):
-
     def get_vocab(self, vocab_type='word'):
+        """Produce a dictionary for the vocab feature name
+
+        :param vocab_type: The feature name
+        """
         return self.word_vocab if vocab_type == 'word' else self.char_vocab
 
-    def get_labels(self):
+    def get_labels(self) -> List[str]:
+        """Get the labels (names of each class)
+
+        :return: (`List[str]`) The labels
+        """
         return self.labels
 
-    def predict(self, batch_dict, **kwargs):
+    def predict(self, batch_dict: Dict[str, TensorDef], **kwargs):
+        """Take in a batch of data, and predict the tags
+
+        :param batch_dict: A batch of features that is to be predicted
+        :param kwargs: See Below
+
+        :Keyword Arguments:
+
+        * *numpy_to_tensor* (``bool``) Should we convert input from numpy to `torch.Tensor` Defaults to `True`
+        :return: A batch-sized list of predictions
+        """
         numpy_to_tensor = bool(kwargs.get('numpy_to_tensor', True))
         inputs, perm_idx = self.make_input(batch_dict, perm=True, numpy_to_tensor=numpy_to_tensor)
         outputs = self(inputs)
         return unsort_batch(outputs, perm_idx)
 
+    @classmethod
+    def create(cls, embeddings: Dict[str, TensorDef], labels: List[str], **kwargs) -> 'TaggerModelBase':
+        """Create a tagger from the inputs.  Most classes shouldnt extend this
+
+        :param embeddings: A dictionary containing the input feature indices
+        :param labels: A list of the labels (tags)
+        :param kwargs: See below
+
+        :Keyword Arguments:
+
+        * *lengths_key* (`str`) Which feature identifies the length of the sequence
+        * *activation* (`str`) What type of activation function to use (defaults to `tanh`)
+        * *dropout* (`str`) What fraction dropout to apply
+        * *dropin* (`str`) A dictionarwith feature keys telling what fraction of word masking to apply to each feature
+
+        :return:
+        """
+        model = cls()
+        model.lengths_key = kwargs.get('lengths_key')
+        model.activation_type = kwargs.get('activation', 'tanh')
+        model.pdrop = float(kwargs.get('dropout', 0.5))
+        model.dropin_values = kwargs.get('dropin', {})
+        model.labels = labels
+        model.create_layers(embeddings, **kwargs)
+        return model
+
+    def create_layers(self, embeddings: Dict[str, TensorDef], **kwargs):
+        """This method defines the model itself, and must be overloaded by derived classes
+
+        This function will update `self` with the layers required to execute the `call()` method
+
+        :param embeddings: The input feature indices
+        :param kwargs:
+        :return:
+        """
+    def compute_loss(self, inputs):
+        """Define a loss function from the inputs, which includes the gold tag values as `inputs['y']`
+
+        :param inputs:
+        :return:
+        """
+
+
+class TransducerTaggerModel(TaggerModelBase):
```

Comment:
I see that in the docstring there is a mention that encode and transduction mean the same thing but it seems confusing to call the class `Transducer` and the method `init_encode` This should probably match?

Reply:
I dont really like calling it a `Transducer` TBH at all, and the steps are divided into `embed`, `encode` and `decode`.  I dont want to make  it seem like its an encoder-decoder -- its not really, but the transduce just comes from a lack of a better name for the encoder portion thats going on -- calling it encode doesnt seem like it makes sense

Reply:
I guess I dont have a great answer for this.  Whats there now minimally changes the existing interface but I think its an improvement.  If we decide on a better name, happy to make a PR for it, but Id rather not block this PR on that unless we truly hate it

Reply:
Changed to `AbstractEncoderTaggerModel`